File: sdk/alcedo-sdk-node/src/kv.ts
export function createKvResource(ky: any) {
  return {
    get: (key: string, options?: any) =>
      ky.get(`kv/${key}`, options).json(),

    set: (key: string, value: any, ttl?: number, options?: any) =>
      ky.put(`kv/${key}`, {
        json: { value: typeof value === "object" ? JSON.stringify(value) : value },
        searchParams: ttl !== undefined ? { ttl: String(ttl) } : undefined,
        ...options,
      }).json(),

    delete: (key: string, options?: any) =>
      ky.delete(`kv/${key}`, options).json(),

    exists: (key: string, options?: any) =>
      ky.get(`kv/${key}/exists`, options).json(),

    ttl: (key: string, options?: any) =>
      ky.get(`kv/${key}/ttl`, options).json(),

    list: (prefix?: string, options?: any) =>
      ky.get("kv", {
        searchParams: prefix !== undefined ? { prefix } : undefined,
        ...options,
      }).json(),

    batch_get: (keys: string[], options?: any) =>
      ky.post("kv/batch/get", { json: { keys }, ...options }).json(),

    batch_set: (pairs: Array<{ key: string; value: any; ttl?: number }>, options?: any) =>
      ky.post("kv/batch/set", { json: pairs, ...options }).json(),

    batch_delete: (keys: string[], options?: any) =>
      ky.post("kv/batch/delete", { json: { keys }, ...options }).json(),

    query: (pattern?: string, options?: any) =>
      ky.get("kv/query", {
        searchParams: pattern !== undefined ? { pattern } : undefined,
        ...options,
      }).json(),
  };
}

File: sdk/alcedocore-sdk-python/alcedo_sdk/resources/kv.py
import json
from typing import Any, Optional

from ..exceptions import NotFoundError, ValidationError
from .base import BaseResource


class KVResource(BaseResource):
    """Key-Value store resource. Maps to /api/kv/* endpoints."""

    async def get(self, key: str) -> Optional[Any]:
        """Get a value by key. Returns None if key doesn't exist."""
        data = await self._request("GET", f"/api/kv/{key}")
        return data.get("data") if data else None

    async def set(self, key: str, value: Any, ttl: Optional[int] = None) -> Any:
        """Set a key-value pair with optional TTL (seconds). Returns the stored value."""
        payload: dict[str, Any] = {"value": value if isinstance(value, (str, bytes)) else json.dumps(value)}
        params = {"ttl": str(ttl)} if ttl is not None else None
        data = await self._request("PUT", f"/api/kv/{key}", json=payload, params=params)
        return data.get("data") if data else None

    async def delete(self, key: str) -> bool:
        """Delete a key. Returns True if deleted, False if key didn't exist."""
        try:
            await self._request("DELETE", f"/api/kv/{key}")
            return True
        except NotFoundError:
            return False

    async def exists(self, key: str) -> bool:
        """Check if a key exists."""
        data = await self._request("GET", f"/api/kv/{key}/exists")
        return data.get("exists", False) if data else False

    async def ttl(self, key: str) -> Optional[int]:
        """Get TTL for a key in seconds. Returns None if no TTL set."""
        data = await self._request("GET", f"/api/kv/{key}/ttl")
        if data is None:
            return None
        val = data.get("ttl")
        return int(val) if val is not None else None

    async def list_keys(self, prefix: str = "") -> list[str]:
        """List all keys, optionally filtered by prefix."""
        params = {"prefix": prefix} if prefix else None
        data = await self._request("GET", "/api/kv/", params=params)
        return data.get("keys", []) if data else []

    async def batch_get(self, keys: list[str]) -> dict[str, Optional[Any]]:
        """Get multiple keys at once. Returns dict of key -> value (None for missing keys)."""
        data = await self._request("POST", "/api/kv/batch/get", json={"keys": keys})
        return data.get("values", {}) if data else {}

    async def batch_set(self, pairs: list[dict[str, Any]]) -> None:
        """Set multiple key-value pairs at once. Each pair: {"key": str, "value": any, "ttl": optional int}."""
        items: list[dict[str, Any]] = []
        for pair in pairs:
            if "key" not in pair:
                raise ValidationError(
                    "Each pair must contain a 'key' field",
                    details={"pair": pair},
                )
            if "value" not in pair:
                raise ValidationError(
                    "Each pair must contain a 'value' field",
                    details={"pair": pair},
                )
            item: dict[str, Any] = {
                "key": pair["key"],
                "value": str(pair["value"]),
            }
            if "ttl" in pair:
                item["ttl"] = pair["ttl"]
            items.append(item)
        await self._request("POST", "/api/kv/batch/set", json=items)

    async def batch_delete(self, keys: list[str]) -> int:
        """Delete multiple keys at once. Returns count of deleted keys."""
        data = await self._request("POST", "/api/kv/batch/delete", json={"keys": keys})
        return data.get("deleted", 0) if data else 0
